File: web/src/lib/utils/inline-media.ts
/**
 * Inline image tokens embedded in entry bodies (markdown image syntax), so
 * photos can sit BETWEEN paragraphs instead of only in the bottom gallery.
 *
 * Form: `![caption](wayli-media:<ref>)` where `<ref>` is whatever the writing
 * platform stores in `trip_media.storage_path` — the full public URL (web
 * uploads) or the raw bucket path (Android uploads, resolved to a signed URL
 * client-side). Renderers strip the `wayli-media:` prefix and turn the ref
 * into a loadable URL before passing the body through the markdown renderer
 * (which already draws `![…](…)` images).
 */

import { fluxbase } from '$lib/fluxbase';

const MEDIA_TOKEN = /!\[([^\]]*)\]\(wayli-media:([^)\s]+)\)/g;

/** Parentheses would end the markdown link early — percent-encode them. */
function encodeRef(ref: string): string {
	return ref.replaceAll('(', '%28').replaceAll(')', '%29');
}

/** Build an inline image token for a stored media reference. */
export function mediaToken(ref: string, caption = 'photo'): string {
	return `![${caption}](wayli-media:${encodeRef(ref)})`;
}

/** Decode a token reference back to the raw storage path / URL form. */
function decodeRef(ref: string): string {
	try {
		return decodeURIComponent(ref);
	} catch {
		return ref;
	}
}

/**
 * Storage references currently placed inline in a body — compare against
 * `trip_media.storage_path` to keep those photos out of the bottom gallery.
 */
export function inlineMediaRefs(body?: string | null): Set<string> {
	if (!body) return new Set();
	const refs = new Set<string>();
	for (const match of body.matchAll(MEDIA_TOKEN)) {
		refs.add(decodeRef(match[2]));
	}
	return refs;
}

/**
 * A raw bucket path becomes its public URL (the trip-images bucket is
 * public-read); absolute URLs pass through unchanged.
 */
export function storageRefToUrl(ref: string): string {
	if (/^https?:\/\//i.test(ref)) return ref;
	const { data } = fluxbase.storage.from('trip-images').getPublicUrl(ref);
	return data.publicUrl;
}

/**
 * Replace every inline token with a plain markdown image whose destination is
 * a loadable URL. Unresolvable refs are dropped rather than rendered broken.
 */
export function resolveInlineMedia(body?: string | null): string {
	if (!body) return '';
	return body.replace(MEDIA_TOKEN, (_m, caption: string, ref: string) => {
		return `![${caption}](${storageRefToUrl(decodeRef(ref))})`;
	});
}

File: android/app/src/main/kotlin/io/github/nimbleflux/wayli/feature/travel/InlineMedia.kt
package io.github.nimbleflux.wayli.feature.travel

/**
 * Inline image tokens embedded in entry bodies (markdown image syntax), so
 * photos can sit BETWEEN paragraphs instead of only in the bottom gallery.
 *
 * Two reference forms:
 * - `![caption](wayli-media:<storagePath>)` — an uploaded photo (server media).
 * - `![photo](wayli-draft:<index>)` — a not-yet-uploaded local pick; the index
 *   points into the editor session's local photo list and is rewritten to the
 *   final `wayli-media:` form when the entry publishes (online or via the
 *   sync worker). This keeps composing fully offline-capable.
 *
 * Both apps' markdown renderers already draw `![…](…)` images; renderers just
 * pre-resolve the `wayli-media:`/`wayli-draft:` destinations to display URLs.
 */
object InlineMedia {
    private val DRAFT_TOKEN = Regex("""!\[([^\]]*)\]\(wayli-draft:(\d+)\)""")
    private val MEDIA_TOKEN = Regex("""!\[([^\]]*)\]\(wayli-media:([^)\s]+)\)""")

    fun draftToken(index: Int): String = "![photo](wayli-draft:$index)"

    /** Append a local-pick token as its own paragraph at the end of the body. */
    fun appendDraftImage(body: String, index: Int): String {
        val token = draftToken(index)
        if (body.isBlank()) return token
        return body.trimEnd() + "\n\n" + token + "\n"
    }

    /**
     * Replace every `wayli-draft:<i>` token with the final `wayli-media:` form
     * using [storagePathByIndex]; tokens whose photo vanished (removed before
     * publish) are dropped.
     */
    fun rewriteDraftTokens(body: String, storagePathByIndex: (Int) -> String?): String =
        DRAFT_TOKEN.replace(body) { m ->
            val idx = m.groupValues[2].toIntOrNull()
            val path = idx?.let(storagePathByIndex)
            if (path != null) "![${m.groupValues[1]}](wayli-media:$path)" else ""
        }

    /** Storage paths referenced by inline media tokens (for gallery filtering). */
    fun inlineMediaPaths(body: String?): Set<String> =
        body?.let { MEDIA_TOKEN.findAll(it).map { m -> m.groupValues[2] }.toSet() } ?: emptySet()

    /** Local-photo indexes referenced by inline draft tokens (editor preview filtering). */
    fun inlineDraftIndexes(body: String?): Set<Int> =
        body?.let { DRAFT_TOKEN.findAll(it).mapNotNull { m -> m.groupValues[2].toIntOrNull() }.toSet() } ?: emptySet()

    /**
     * Turn tokens into renderable markdown: `wayli-media:` destinations go
     * through [resolveMedia] (signed/public URL by storage path), draft tokens
     * through [resolveDraft] (local file URL by index). Unresolvable images are
     * dropped rather than rendered as broken placeholders.
     */
    fun resolve(body: String, resolveMedia: (String) -> String?, resolveDraft: (Int) -> String? = { null }): String {
        var resolved = MEDIA_TOKEN.replace(body) { m ->
            resolveMedia(m.groupValues[2])?.let { url -> "![${m.groupValues[1]}]($url)" } ?: ""
        }
        resolved = DRAFT_TOKEN.replace(resolved) { m ->
            val idx = m.groupValues[2].toIntOrNull()
            idx?.let(resolveDraft)?.let { url -> "![${m.groupValues[1]}]($url)" } ?: ""
        }
        return resolved
    }
}
